packages/react-aria-components/src/HiddenDateInput.tsx

```diff
@@ -0,0 +1,135 @@
+  /**
+   * State for the input.
+   */
+  state: DateFieldState | DatePickerState
+}
+
+export interface HiddenDateAria {
+  /** Props for the container element. */
+  containerProps: React.HTMLAttributes<HTMLDivElement>,
+  /** Props for the hidden input element. */
+  inputProps: React.InputHTMLAttributes<HTMLInputElement>
+}
+
+const dateSegments = ['day', 'month', 'year'];
+const granularityMap = {'hour': 1, 'minute': 2, 'second': 3};
+
+export function useHiddenDateInput(props: HiddenDateInputProps, state: DateFieldState | DatePickerState) : HiddenDateAria {
+  let {
+    autoComplete,
+    isDisabled,
+    name
+  } = props;
+  let {visuallyHiddenProps} = useVisuallyHidden();
+
+  let inputStep = 60;
+  if (state.granularity === 'second') {
+    inputStep = 1;
+  } else if (state.granularity === 'hour') {
+    inputStep = 3600; 
+  }
+
+  let dateValue = state.value == null ? '' : state.value.toString();
+
+  let inputType = state.granularity === 'day' ? 'date' : 'datetime-local';
+
+  let timeSegments = ['hour', 'minute', 'second'];
+  // Depending on the granularity, we only want to validate certain time segments
+  let end = 0;
+  if (timeSegments.includes(state.granularity)) {
+    end = granularityMap[state.granularity];
+    timeSegments = timeSegments.slice(0, end);
+  }
+
+  return {
+    containerProps: {
+      ...visuallyHiddenProps,
+      'aria-hidden': true,
+      // @ts-ignore
+      ['data-react-aria-prevent-focus']: true,
+      // @ts-ignore
+      ['data-a11y-ignore']: 'aria-hidden-focus'
+    },
+    inputProps: {
+      tabIndex: -1,
+      autoComplete,
+      disabled: isDisabled,
+      type: inputType,
+      // We set the form prop to an empty string to prevent the hidden date input's value from being submitted
+      form: '',
+      name,
+      step: inputStep,
+      value: dateValue,
+      onChange: (e) => {
+        let targetString = e.target.value.toString();
+        if (targetString) {
+          try {
+            let targetValue: CalendarDateTime | CalendarDate = parseDateTime(targetString);
+            if (state.granularity === 'day') {
+              targetValue = parseDate(targetString);
+            }
+            // We check to to see if setSegment exists in the state since it only exists in DateFieldState and not DatePickerState.
+            // The setValue method has different behavior depending on if it's coming from DateFieldState or DatePickerState.
+            // In DateFieldState, setValue firsts checks to make sure that each segment is filled before committing the newValue 
+            // which is why in the code below we first set each segment to validate it before committing the new value. 
+            // However, in DatePickerState, since we have to be able to commit values from the Calendar popover, we are also able to 
+            // set a new value when the field itself is empty.
+            if ('setSegment' in state) {
```

Comment:
We could probably make `state.setValue` a bit smarter so it sets all the segments as valid when called from the outside. We can do that later.